### PR TITLE
std.format: Add description of flags, width, precision, separator and positional arguments.

### PR DESCRIPTION
**Pull request by berni44**
I kept the description of the flags rather brief. They will be given in more detail in the followup PR, which will feature a large table of types and format chars, which will give special meanings of the flags in more detail.

### CODE REVIEW

File: std/format/package.d [hunk 1]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
+             right justified. With the $(B '-') flag this is changed
+             to left justification.
+
+             There are two exceptions where the $(B '-') flag has a
+             different meaning: With $(B 'r') it denotes to use little
+             endian and in case of a compound indicator it means that
+             no special handling of the members is applied.))
+   $(TR $(TD $(B '+')&nbsp;/&nbsp;$(B '&nbsp;'))
+        $(TD Applies only to numbers. Normally, positiv numbers are not
+             preceded by anything. With one of these two flags present,
```
Comment (RazvanN7) [inline std/format/package.d:219]:
```suggestion
             formatted to include the $(B +) sign. With one of these two flags present,
```

File: std/format/package.d [hunk 2]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
+             positiv numbers are preceded by a plus sign or a space.
```
Comment (RazvanN7) [inline std/format/package.d:220]:
```suggestion
             positive numbers are preceded by a plus sign or a space.
```

File: std/format/package.d [hunk 3]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
+             When both flags are present, a plus sign is used.
+
+             In case of $(B 'r') it denotes to use big endian.))
```
Comment (RazvanN7) [inline std/format/package.d:223]:
```suggestion
             In case of $(B 'r'), a big endian format is used.))
```

File: std/format/package.d [hunk 4]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
+   $(TR $(TD $(B '0'))
+        $(TD Applies only to numbers that are printed right justified.
```
Comment (RazvanN7) [inline std/format/package.d:225]:
```suggestion
        $(TD Is applied to numerical values that are printed right justified.
```

File: std/format/package.d [hunk 5]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
+             If the zero flag is present, the space left to the number is
+             filled with zeros instead of spaces.))
+   $(TR $(TD $(B '#'))
+        $(TD An alternativ output shall be used. This depends on the type
```
Comment (RazvanN7) [inline std/format/package.d:229]:
```suggestion
        $(TD Denotes that an alternative output must be used. This depends on the type
```

File: std/format/package.d [hunk 6]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
+             sections below for more information.))
+)
 
-    $(BOOKTABLE Flags affect formatting depending on the specifier as
-    follows., $(TR $(TH Flag) $(TH Types&nbsp;affected) $(TH Semantics))
+$(SECTION4 Width\, Precision and Separator)
 
-    $(TR $(TD $(B '-')) $(TD numeric, bool, null, char, string, enum, pointer) $(TD Left justify the result in
-        the field.  It overrides any $(B 0) flag.))
+The $(I width) specifies the minimum width of the result.
```
Comment (RazvanN7) [inline std/format/package.d:236]:
```suggestion
The $(I width) parameter specifies the minimum width of the result.
```

Comment (RazvanN7) [reply std/format/package.d:236]: This sentence is kind of tautological: "The width specifies the width of the result".

File: std/format/package.d [hunk 7]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
-    formatting is a '0', even if both the argument and the $(I
-    Precision) are zero.))
+A $(I separator) is only used for formatting numbers. If it is
+specified, the output is divided into chunks of three digits,
+separated by a $(B ','), for better readability. The number of digits
+in a chunk can also be given explicitly.
 
-    $(TR $(TD $(B '#')) $(TD integral ($(B 'x'), $(B 'X'))) $(TD If
-       non-zero, prefix result with $(B 0x) ($(B 0X)).))
+In all three cases the number can be replaced by a $(B '*'). In that
```
Comment (RazvanN7) [inline std/format/package.d:249]:
```suggestion
In all three cases the number of digits can be replaced by a $(B '*').
```

File: std/format/package.d [hunk 8]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
+case the next argument is used as the number. If the argument is a
```
Comment (RazvanN7) [inline std/format/package.d:250]:
```suggestion
In this scenario, the next argument is used as the number of digits. If the argument is a
```

File: std/format/package.d [hunk 9]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
+negative number, in case of $(I width), the $(B '-') flag is applied,
```
Comment (RazvanN7) [inline std/format/package.d:251]:
```suggestion
negative number,  the $(I width) parameter is set to $(B '-'),
```

File: std/format/package.d [hunk 10]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
+in case of $(I precision) and $(I separator) it is, as if no
+precision or separator is specified.
```
Comment (RazvanN7) [inline std/format/package.d:253]: Replace the above two lines with: " whereas the $(I precision) and $(I separator) parameters are considered unspecified". Overall, the paragraph will look something like this:

"In all three cases the number of digits can be replaced by a $(B '*'). In this scenario, the next argument is used as the number of digits. If the argument is a negative number,  the $(I width) parameter is set to $(B '-'), whereas the $(I precision) and $(I separator) parameters are considered unspecified."

I hope I did not change the semantical meaning of the paragraph. What do you think?

File: std/format/package.d [hunk 11]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
 
-    $(TR $(TD $(B '#')) $(TD floating) $(TD Always insert the decimal
-       point and print trailing zeros.))
+The $(I separator) can also be followed by a $(B '?'). In that case
```
Comment (RazvanN7) [inline std/format/package.d:255]:
```suggestion
The $(I separator) can also be followed by a $(B '?'). In that case,
```

File: std/format/package.d [hunk 12]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
 
-    $(TR $(TD $(B '0')) $(TD numeric) $(TD Use leading
-    zeros to pad rather than spaces (except for the floating point
-    values `nan` and `infinity`).  Ignore if there's a $(I
-    Precision).))
+$(SECTION4 Position)
 
-    $(TR $(TD $(B ' ')) $(TD numeric) $(TD Prefix positive
-    numbers in a signed conversion with a space.)))
+Normally, the arguments are processed in order. With the $(I
```
Comment (RazvanN7) [inline std/format/package.d:261]:
```suggestion
By default, the arguments are processed in order. With the $(I
```

Comment (RazvanN7) [reply std/format/package.d:261]: What kind of order do you refer to? Lexical order?

File: std/format/package.d [hunk 13]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
-        The fractional part of floating point values inserts the separator
-        from left to right.
-        Entering an integer after the ',' allows to specify $(I X).
-        If a '*' is placed after the ',' then $(I X) is specified by an
-        additional parameter to the format function.
-        Adding a '?' after the ',' or $(I X) specifier allows to specify
-        the separator character as an additional parameter.
-        )
+It's also possible to use positional arguments for $(I width), $(I
+precision) and $(I separator), by adding a number and a $(B
```
Comment (RazvanN7) [inline std/format/package.d:269]:
```suggestion
precision) and $(I separator) by adding a number and a $(B
```

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
A couple of suggestions for future PRs, if I may:

 - use "by default" instead of "normally"
 - try to not repeat the same expression in one paragraph, but rather use synonyms. For example: you use "in this case" a lot. There were paragraphs of 5 lines that had "in this case" 6-7 times. These can be  substituted by "in this scenario", "in this situation", "in these circumstances" etc. or could be ditched entirely by carefully rephrasing the sentence.
 - aim for small and concise sentences.

File: std/format/package.d (hunk 6 above)
Comment (berni44) [reply std/format/package.d:236]: The important thing is _minimum_.

File: std/format/package.d (hunk 12 above)
Comment (berni44) [reply std/format/package.d:261]: From left to right (the order given in the function call).

File: std/format/package.d (hunk 10 above)
Comment (berni44) [reply std/format/package.d:253]: I think everything is ok, but the width stuff. The width parameter is set to `abs(value)` and the dash-flag is set. 

I rewrote it this way:

"If the argument is a negative number, the $(I precision) and
$(I separator) parameters are considered unspecified. For ($I width)
the absolute value is used and the $(B '-') flag is set."

OK?

**Comment by berni44** [comment]
> A couple of suggestions for future PRs, if I may:

I'll try. :-)

> * try to not repeat the same expression in one paragraph, but rather use synonyms.

That's not easy for me. I'm often happy when I manage to find one way to express what I want to say. Sometimes I find synonyms in the dictionary, but I'm unsure if they really express, what I want to say. For example, I wanted to write here again "I'll try", like above. The dictionary suggest "I'll struggle", but I've got the feeling, that this is something else. Same with "I'll bother". Maybe "I'll make an effort" might be good?

Or turning it the other way round: I'm quite thankful, that you are helping me here... :-)

File: std/format/package.d [hunk 14]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
+that are included in the output.
 
-    $(TR $(TD $(B '#')) $(TD integral ($(B 'o'))) $(TD Add to
-    precision as necessary so that the first digit of the octal
-    formatting is a '0', even if both the argument and the $(I
-    Precision) are zero.))
+A $(I separator) is only used for formatting numbers. If it is
+specified, the output is divided into chunks of three digits,
+separated by a $(B ','), for better readability. The number of digits
+in a chunk can also be given explicitly.
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply std/format/package.d:247]: I suggest specifying how you can do that. E.g.: "The number of digits in a chunk can be given explicitly by [...]"

File: std/format/package.d [hunk 15]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
 
-    $(TR $(TD $(B '#')) $(TD integral ($(B 'x'), $(B 'X'))) $(TD If
-       non-zero, prefix result with $(B 0x) ($(B 0X)).))
+In all three cases the number of digits can be replaced by a $(B
+'*'). In this scenario, the next argument is used as the number of
+digits. If the argument is a negative number, the $(I precision) and
+$(I separator) parameters are considered unspecified. For ($I width)
```
Comment (RazvanN7) [inline std/format/package.d:252]:
```suggestion
$(I separator) parameters are considered unspecified. For ($I width),
```

File: std/format/package.d [hunk 16]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
+the absolute value is used and the $(B '-') flag is set.
```
Comment (RazvanN7) [inline std/format/package.d:253]: Nice, this last sentence sounds so much better.

File: std/format/package.d [hunk 17]
```diff
@@ -198,66 +198,81 @@ contain exactly one $(I format specifier), which specifies the way,
...
 
-    $(TR $(TD $(B '0')) $(TD numeric) $(TD Use leading
-    zeros to pad rather than spaces (except for the floating point
-    values `nan` and `infinity`).  Ignore if there's a $(I
-    Precision).))
+$(SECTION4 Position)
 
-    $(TR $(TD $(B ' ')) $(TD numeric) $(TD Prefix positive
-    numbers in a signed conversion with a space.)))
+By default, the arguments are processed in order. With the $(I
```
Comment (RazvanN7) [inline std/format/package.d:261]:
```suggestion
By default, the arguments are processed in the provided order. With the $(I
```

File: std/format/package.d (hunk 10 above)
Comment (RazvanN7) [reply std/format/package.d]: Yeah, that's great thanks!

**Comment by RazvanN7** [comment]
> Or turning it the other way round: I'm quite thankful, that you are helping me here... :-)

Don't worry about it, it's my pleasure.

File: std/format/package.d (hunk 14 above)
[the remark this answers is not in the document]
Comment (berni44) [reply std/format/package.d:247]: "by providing a number or a $(B '*') after the $(B ',')." ?

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
Looks good! Thanks! Might need a rebase though if buildkite keeps failing.

**Comment by thewilsonator** [comment]
`_D3std6format8internal6floats__T11printFloatFVbi1TdTaZQwFNaNfNkAadSQCnQCm4spec__T10FormatSpecTaZQpEQDtQDsQDoQDi12RoundingModeAyaimbZQCr: error: undefined reference to '_D3std4math4ceilFNaNbNiNeeZe'`

**Comment by RazvanN7** [comment]
Obviously not related to this PR (which simply adds documentation)

**Comment by berni44** [comment]
> `_D3std6format8internal6floats__T11printFloatFVbi1TdTaZQwFNaNfNkAadSQCnQCm4spec__T10FormatSpecTaZQpEQDtQDsQDoQDi12RoundingModeAyaimbZQCr: error: undefined reference to '_D3std4math4ceilFNaNbNiNeeZe'`

IMHO this is a bug in the testers. They mix different versions of Phobos whenever the branch is not based on the HEAD. Sometimes this leads to such messages. We had several of them in the past, when we split `std.format`. So this is probably due to the merging of PR #7956 between creating this PR and now. I'll rebase and it will vanish.

**Pull request merged**